helper/TypeMapping.js

```diff
@@ -0,0 +1,146 @@
+const _ = require('lodash');
+const elasticsearch = require('elasticsearch');
+
+var TypeMapping = function(){
+
+  // A list of all sources
+  this.sources = [];
+
+  // A list of alternate names for sources, mostly used to save typing
+  this.source_aliases = {};
+
+  // A list of all layers
+  this.layers = [];
+
+  /*
+   * A list of all layers in each source. This is used for convenience elswhere
+   * and to determine when a combination of source and layer parameters is
+   * not going to match any records and will return no results.
+   */
+  this.layers_by_source = {};
+
+  /*
+   * A list of layer aliases that can be used to support specific use cases
+   * (like coarse geocoding) * or work around the fact that different sources
+   * may have layers that mean the same thing but have a different name
+   */
+  this.layer_aliases = {};
+
+  /*
+   * An object that contains all sources or aliases. The key is the source or alias,
+   * the value is either that source, or the canonical name for that alias if it's an alias.
+   */
+  this.source_mapping = {};
+
+  /*
+   * An object that has a key for each possible layer or alias,
+   * and returns either that layer, or gall the layers in the alias
+   */
+  this.layer_mapping = {};
+};
+
+TypeMapping.addStandardTargetsToAliases = function(standard, aliases) {
+  var combined = _.extend({}, aliases);
+  standard.forEach(function(target) {
+    if (combined[target] === undefined) {
+      combined[target] = [target];
+    }
+  });
+
+  return combined;
+};
+
+// source alias setter
+TypeMapping.prototype.setSourceAliases = function( aliases ){
+  this.source_aliases = aliases;
+};
+
+// layers-by-source alias setter
+TypeMapping.prototype.setLayersBySource = function( lbs ){
+  this.layers_by_source = lbs;
+};
+
+// layer alias setter
+TypeMapping.prototype.setLayerAliases = function( aliases ){
+  this.layer_aliases = aliases;
+};
+
+// generate dynamic mappings after setters have been run
+TypeMapping.prototype.generateDynamicMappings = function(){
```

Comment:
This shadows an Elasticsearch term which means something pretty different, maybe we can come up with a different name?

Reply:
I'm happy to change it, please suggest an alternative.